### PR TITLE
Move Enum and MdImport to shared

### PR DESCRIPTION
**Pull request by EgorBo**
Also, I replaced half-native `CompareTo` with managed version from CoreRT, is it ok or I should attach benchmarks? (if it looks good - I'll delete the native impl for InternalCompare)
CoreRT implementation uses `EETypePtr` here and there in `Enum` so it's not easy to reuse this implementation (I managed to extract ~40% of common code between CoreCLR and CoreRT and 95% between CoreCLR and Mono).

Naive [Enum.Mono.cs](https://gist.github.com/EgorBo/3a014bbaea492a36e4713d20252156f7) implementation.

/cc @marek-safar @jkotas

### CODE REVIEW

**Comment by jkotas** [comment]
> is it ok or I should attach benchmarks? 

Yes, benchmarks to double check would be good.

**Comment by jkotas** [comment]
Note that the Enum implementation overall is not particularly great. https://github.com/dotnet/coreclr/pull/22161 has good ideas how to make it better, but it uses generics too much. Once we get to sorting this out, I expect that it will change a lot.

**Comment by EgorBo** [comment]
From a quick glance that implementation looks mono-friendly

**Comment by jkotas** [comment]
> From a quick glance that implementation looks mono-friendly

But I expect that the ties (FCalls, etc.) it is going to have with the rest of the runtime will be very different.

File: src/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 1]
```diff
@@ -204,7 +204,6 @@
     <Compile Include="$(BclSourcesRoot)\System\Reflection\LoaderAllocator.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Reflection\MdConstant.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Reflection\MdFieldInfo.cs" />
-    <Compile Include="$(BclSourcesRoot)\System\Reflection\MdImport.cs" />
```
Comment (jkotas) [inline src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Does Enum depend on MdImport anywhere, or is that unrelated change?

Comment (EgorBo) [reply src/System.Private.CoreLib/System.Private.CoreLib.csproj]: @jkotas Enum uses CorElementType from that file, I guess other types in that file will be useful too soon.

Comment (EgorBo) [reply src/System.Private.CoreLib/System.Private.CoreLib.csproj]: CorElementType in mono is a little bit different,e.g. `CorElementType.Boolean` instead of `CorElementType.ELEMENT_TYPE_BOOLEAN` in CoreCLR.

**Comment by MichalStrehovsky** [comment]
I thought we don't put methods marked InternalCall into shared partition.

Also `RuntimeType` is not a shared type and we haven't been using it in the shared partition (except for some odd `.RuntimeType.cs` dot files).

**Comment by EgorBo** [comment]
@MichalStrehovsky I remove the internalcalls from the shared part but I can't remove RuntimeType stuff without it I won't be able to share 95% of code with mono.
@jkotas I benchmarked CompareTo using the following benchmark:
```csharp
using System;
using System.Collections.Generic;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace ConsoleApp134
{
    public class Program
    {
        static void Main(string[] args)
        {
            BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run();
        }

        [Benchmark]
        [ArgumentsSource(nameof(TestData))]
        public int CompareEnums(Enum left, object right, string description)
        {
            try
            {
                return left.CompareTo(right);
            }
            catch
            {
                return 42;
            }
        }

        public IEnumerable<object[]> TestData()
        {
            yield return new object[] { UInt8Enum.c, UInt8Enum.c, "case 1" }; // same types, same values
            yield return new object[] { UInt8Enum.c, UInt8Enum.d, "case 2" }; // same types, different values
            yield return new object[] { UInt8Enum.c, Int64Enum.a, "case 3" }; // different types (throws an exception) 

            yield return new object[] { Int64Enum.c, Int64Enum.c,  "case 4" }; // same types, same values
            yield return new object[] { Int64Enum.c, Int64Enum.d,  "case 5" }; // same types, different values
            yield return new object[] { Int64Enum.c, Uint32Enum.a, "case 6" }; // different types (throws an exception) 
        }
    }

    // some random enums

    [Flags]
    public enum UInt8Enum : byte
    {
        a = 0x00,
        b = 0x01,
        c = 0x02,
        d = 0x04,
        e = 0x08,
        f = 0x10,
        g = 0x20,
        x = 0x40,
    }

    public enum Int64Enum : long
    {
        a,
        b,
        c,
        d,
        e
    }


    public enum Uint32Enum : uint
    {
        a,
        b,
        c,
        d,
        e,
        f,
    }

    public enum DefauleEnum
    {
        a,
        b,
        c,
        d
    }
}
```
The difference is ~+/-1% between my branch and master.

**Comment by EgorBo** [comment]
@jkotas 
It seems the the managed impl of CompareTo is slower
Benchmark - Enum.CompareTo.Benchmark.cs
Results:
```
| description |           New |           Old |
|------------ |--------------:|--------------:|
|      case 1 |      9.908 ns |      6.367 ns |
|      case 2 |      9.875 ns |      6.371 ns |
|      case 3 | 12,054.080 ns | 12,028.336 ns |
|      case 4 |     10.137 ns |      6.756 ns |
|      case 5 |     10.135 ns |      6.529 ns |
|      case 6 | 12,032.973 ns | 11,741.258 ns |
```
New - my branch
Old - master

**Comment by jkotas** [comment]
That is not that surprising given how it is implemented. I do not think we want to take this regression.

**Comment by EgorBo** [comment]
@jkotas yeah I undid that change

File: src/System.Private.CoreLib/System.Private.CoreLib.csproj (hunk 1 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/System.Private.CoreLib.csproj]: @EgorBo extract CorElementType into its own file for now and keep the rest there

File: src/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    partial class Enum
```
Comment (marek-safar) [inline src/System.Private.CoreLib/src/System/Enum.CoreCLR.cs]: CoreCLR prefers to duplicate everything, it should be `public abstract partial class Enum : ValueType, IComparable, IFormattable, IConvertible`

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Enum.CoreCLR.cs]: We prefer to duplicate the modifiers (public, etc.). I do not think we generally duplicate the list interfaces unless it helps with readability.

**Comment by filipnavara** [comment]
> I can't remove RuntimeType stuff without it I won't be able to share 95% of code with mono.

You can follow the pattern taken with `Activator.RuntimeType.cs` - ie. split the part using `RuntimeType` into separate file in shared partition and guard it appropriately in the shared project.

**Comment by jkotas** [comment]
> You can follow the pattern taken with `Activator.RuntimeType.cs` - ie. split the part using RuntimeType into separate file in shared partition and guard it appropriately in the shared project.

Thoughts about this?

**Comment by marek-safar** [comment]
> Thoughts about this?

Improving CoreRT sharing was nice to have and not required so it can be done later, right?

**Comment by jkotas** [comment]
Right, but should we use a consistent patterns for this?

**Comment by MichalStrehovsky** [comment]
> Improving CoreRT sharing was nice to have and not required so it can be done later, right?

Both files would stay excluded for CoreRT, but this is a structure (putting things requiring RuntimeType into separate dot files) we already have in place. Doing this later just messes up git history for the individual lines. IMHO it's better to do the break in one go.

**Comment by marek-safar** [comment]
What will be in shared can be shared with CoreRT too, probably not much but code like https://github.com/dotnet/coreclr/blob/56d9776ae74e69f6c67f8f234bef85e6180cfa35/src/System.Private.CoreLib/shared/System/Enum.cs#L320 is not fully Enum.RuntimeType.cs specific. We could do the Enum.cs vs Enum.RuntimeType.cs split if that's something you see useful.

**Comment by MichalStrehovsky** [comment]
I did a search and replace of `RuntimeType` to `Type` in this file to get a quick sense of how much this code actually relies on `RuntimeType`. The only method that actually accesses members is `GetCachedValuesAndNames`.

How would people feel if we just replace `RuntimeType` with `Type` and use the cross-runtime portable `IsRuntimeImplemented` method for argument validation. We would cast to `RuntimeType` at boundaries? We used that pattern elsewhere.

**Comment by EgorBo** [comment]
@jkotas @MichalStrehovsky I tried to share some non-RuntimeType related stuff with CoreRT in my recent commits (tested on CoreRT locally).

**Comment by MichalStrehovsky** [comment]
> I tried to share some non-RuntimeType related stuff with CoreRT

How about the replacement of `RuntimeType` with `Type` I suggested above? Is it problematic? It looks like we could move even more to the "shared with everything" part (to the point where the RuntimeType.cs file might no longer be needed). The parsing code is one of the most actively worked on parts, so having it in the right long-term place would be nice.

**Comment by jkotas** [comment]
You may be able to do this by doing `using Runtime = Type;` to keep CoreCLR on the RuntimeType plan if it is important.

**Comment by EgorBo** [comment]
@jkotas @MichalStrehovsky done! Here is the [Enum.CoreRT.cs](https://gist.github.com/EgorBo/26aad61742fcf0d96daefc69347bd1ee) I tested with these changes.
Also, had to change Type a bit (the Enum-related methods) and EnumInfo.

File: src/System.Private.CoreLib/shared/System/Enum.cs [hunk 3]
```diff
@@ -288,94 +159,33 @@ private static string InternalFlagsFormat(RuntimeType eT, TypeValuesAndNames ent
...
-            Debug.Assert(success);
-            return result;
+            // Get all of the names
+            return GetCachedValuesAndNames(enumType, true).Names;
         }
 
-        public static bool TryParse(Type enumType, string value, out object result) =>
-            TryParse(enumType, value, ignoreCase: false, out result);
-
-        public static bool TryParse(Type enumType, string value, bool ignoreCase, out object result) =>
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Enum.cs]: Is there a intention behind some of this reordering? I think the code was much easier to understand when all Parse methods were together.

File: src/System.Private.CoreLib/shared/System/Enum.cs [hunk 4]
```diff
@@ -288,94 +159,33 @@ private static string InternalFlagsFormat(RuntimeType eT, TypeValuesAndNames ent
...
         }
 
-        internal static ulong ToUInt64(object value)
+        internal static string GetEnumName(RuntimeType eT, ulong ulValue)
         {
-            // Helper function to silently convert the value to UInt64 from the other base types for enum without throwing an exception.
-            // This is need since the Convert functions do overflow checks.
-            TypeCode typeCode = Convert.GetTypeCode(value);
+            Debug.Assert(eT != null);
+            ulong[] ulValues = Enum.InternalGetValues(eT);
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Enum.cs]: Nit: `Enum.` is not needed here. We are inside Enum type.

File: src/System.Private.CoreLib/shared/System/Enum.cs [hunk 5]
```diff
@@ -1166,64 +1150,27 @@ public int CompareTo(object target)
...
-            }
+    #endregion
 
-            return InternalHasFlag(flag);
-        }
-
-        #endregion
-
-        #region IConvertable
+    #region IConvertable
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Enum.cs]: IConvertible

**Comment by jkotas** [comment]
> Here is the Enum.CoreRT.cs I tested with these changes. Also, had to change Type a bit (the Enum-related methods) and EnumInfo.

Could you please share link to CoreRT commit or send WIP PR to CoreRT repo that we will cherry pick once the change is mirrored?

**Comment by EgorBo** [comment]
@jkotas ok, so here is the PR https://github.com/dotnet/corert/pull/7180 (looks green, the failing lanes as far as I understand are not related)
they only thing I am not sure is how to properly implement `GetCachedValuesAndNames` for CoreRT, CoreCLR has `GenericCache` property in RuntimeType

**Comment by jkotas** [comment]
`TypeValuesAndNames` in CoreCLR is the same thing as `EnumInfo` in CoreRT. `EnumInfo` is cached on the runtime type same way as `TypeValuesAndNames` is cached on the runtime type in CoreCLR.

So the implement the caching in CoreRT, make `EnumInfo` look like `TypeValuesAndNames`.

**Comment by jkotas** [comment]
Thanks! Let me know if you are going to finish the CoreRT side. If not, I can take it from here.
